lib/Dialects/LinalgExt/Transforms/TilingToTileOp.cpp

```diff
@@ -52,101 +59,49 @@ static SmallVector<Value> tileToTileOp(PatternRewriter &rewriter,
         SmallVector<Operation *> tiledOps = op.getTiledImplementation(
             b, outSlices, tiledOffsets, tiledSizes, /*tileDestOperands=*/false);
         assert(tiledOps.size() == 1 && "expected single tiled op");
-        Operation *tiledOp = tiledOps.front();
+        tiledOp = tiledOps.front();
         b.create<linalg_ext::TileYieldOp>(loc, tiledOp->getResults());
       });
-  return tileOp->getResults();
+  return TilingResult{tileOp, tiledOp};
 }
 
-namespace {
-
-struct OpTilingPattern : public OpInterfaceRewritePattern<TilingInterface> {
-  OpTilingPattern(MLIRContext *context, linalg::LinalgTilingOptions opt)
-      : OpInterfaceRewritePattern<TilingInterface>(context), options(opt) {}
-
-  LogicalResult matchAndRewrite(TilingInterface op,
-                                PatternRewriter &rewriter) const override {
-    // Poor man's single pattern application to avoid infinite pattern
-    // application wihtout having to hack with attributes and filters.
-    if (remainingApplicationCount <= 0)
-      return failure();
-
-    /// Currently only handle single result operations.
-    if (op->getNumResults() != 1)
-      return failure();
-
-    // Get rank and tile sizes.
-    SmallVector<Value> tileSizes =
-        options.tileSizeComputationFunction(rewriter, op);
-    int64_t dim = -1;
-    for (auto en : llvm::enumerate(tileSizes)) {
-      Optional<int64_t> maybeTileSize = getConstantIntValue(en.value());
-      if (maybeTileSize && *maybeTileSize == 0)
-        continue;
-      if (maybeTileSize && *maybeTileSize < 0)
-        return rewriter.notifyMatchFailure(op, "Negative tile size");
-      if (dim >= 0)
-        return rewriter.notifyMatchFailure(
-            op, "Could not find a single tiling dim");
-      dim = en.index();
-    }
-    if (dim < 0)
+FailureOr<Operation *>
+mlir::linalg_ext::LinalgExtTilingPattern::returningMatchAndRewrite(
+    TilingInterface op, PatternRewriter &rewriter) const {
+  /// Currently only handle single result operations.
+  if (op->getNumResults() != 1)
+    return rewriter.notifyMatchFailure(op, "Not a single result");
+
+  // Get rank and tile sizes.
+  SmallVector<Value> tileSizes =
+      options.tileSizeComputationFunction(rewriter, op);
+  int64_t dim = -1;
+  for (auto en : llvm::enumerate(tileSizes)) {
+    Optional<int64_t> maybeTileSize = getConstantIntValue(en.value());
+    if (maybeTileSize && *maybeTileSize == 0)
+      continue;
+    if (maybeTileSize && *maybeTileSize < 0)
+      return rewriter.notifyMatchFailure(op, "Negative tile size");
+    if (dim >= 0)
```

Comment:
sorry the comment should be here. This checks may be implemented by a transform op verifier.

Reply:
Right there are a few tradeoffs wrt matchers vs transform op verifier vs matchAndRewrite.
I'll add a TODO because I do not know what the end state should look like; if we put everything in the verifier, it will be missing here.

@Mogball in case something comes to mind.

Reply:
the existing ops duplicate some of that verification logic since the patterns are not exclusively used via transform dialect. A comment/TODO is fine...